[Concurrency] Fix task switch performance issue.

al45tair opened the pull request:
When using the new custom default executors, sometimes we end up taking a long time to do a task switch.  This is happening because the path the new code takes sometimes results in a concrete pointer to the default global executor being in the executor tracking information in `swift_task_switch()`, and if we try to switch to a `nil` task executor (which _also_ means the default global executor), we aren’t spotting that and we’re taking the slow path.

Essentially, we want to take the fast path in cases where we switch from `nil` to the concrete default global executor and vice-versa.

rdar://156701386

al45tair committed:
[Concurrency] Fix task switch performance issue.

When using the new custom default executors, sometimes we end up
taking a long time to do a task switch.  This is happening because
the path the new code takes sometimes results in a concrete pointer
to the default global executor being in the executor tracking
information in `swift_task_switch()`, and if we try to switch to
a `nil` task executor (which _also_ means the default global executor),
we aren’t spotting that and we’re taking the slow path.

Essentially, we want to take the fast path in cases where we switch
from `nil` to the concrete default global executor and vice-versa.

rdar://156701386

ktoso reviewed (commented):
Thanks for the description what this change achieves, that does make sense.

I'm wondering if we could write some tests for _just_ the C function `mustSwitchToRun` since we could have this nicely covered when we're expected to switch... but now it's a bit untested and could regress easily 🤔

ktoso reviewed (commented):
How about we add an unit test in `unittests`, similar to how `unittests/Threading/`  tests are set up and check for a bunch of serial/task executors passed to this func? This way we won't accidentally regress functionality here

al45tair committed:
[Concurrency] Further improve performance.

Remove some reference counting traffic by using `Unowned*Executor`s.

Also, add a test to make sure we stay on the fast path.

rdar://156701386

eeckstein committed:
embedded: don't try to specialize witness tables for abstract conformances

If the conformance is abstract the witness table is specialized elsewhere - at the place where the concrete conformance is referenced.
Fixes a compiler crash.

FranzBusch commented:
@swift-ci please build toolchain Ubuntu 24.04

FranzBusch commented:
I just tested this branch locally and it produces the expected performance wins. A custom default executor is now capable of achieving nearly the identical performance of the built-in dispatch global executor.

al45tair commented:
(Cherry picked https://github.com/swiftlang/swift/pull/83391 here so we can test this).

al45tair committed:
[Test][Concurrency] Add libdispatch to the requirements for the test.

The test uses Dispatch, mainly for convenience, but that means it
won't run everywhere.

rdar://156701386

al45tair committed:
[Embedded] Add swift_getDefaultExecutor() to the list of functions to link.

We need to link `swift_getDefaultExecutor`, which is used from the C++
code.

rdar://156701386

al45tair commented:
@swift-ci Please smoke test

ktoso reviewed (approved):
Thanks for adding a test, looks good!
